File: pkg/tree/trees.go
package tree

import (
	"encoding/json"
	"encoding/csv"
	"os"
	"io/ioutil"
	"log"
	"math"
)

type Candidate struct {
	Level   string  `json:"level"`
	Lang    string	`json:"lang"`
	Tweets  bool	`json:"tweets"`
	PhD     bool	`json:"phd"`
	DidWell bool	`json:"didwell"`
}

func (c *Candidate) Get(attr string) interface{} {
	switch attr {
	case "level":
		return c.Level
	case "lang":
		return c.Lang
	case "tweets":
		return c.Tweets
	case "phd":
		return c.PhD
	case "didwell":
		return c.DidWell
	default:
		return nil
	}
}

type Partition map[interface{}][]*Candidate

func fromJSON(jsnfile string) []*Candidate {
	var cands []*Candidate
	data, err := ioutil.ReadFile(jsnfile)
	if err != nil {
		log.Fatalf("cannot read file %s, error: %v\n", jsnfile, err)
	}
	err = json.Unmarshal(data, &cands)
	if err != nil {
		log.Fatal(err)
	}
	return cands
}

func FromCSV(csvfile string) []*Candidate {
	var cands []*Candidate
	file, err := os.Open(csvfile)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	for _, record := range records[1:] {
		cand := &Candidate{Level: record[0], Lang: record[1]}
		if record[2] == "1" {
			cand.Tweets = true
		}
		if record[3] == "1" {
			cand.PhD = true
		}
		if record[4] == "1" {
			cand.DidWell = true
		}
		cands = append(cands, cand)
	}
	return cands
}

func clsprobs(data []interface{}) []float64 {
	total := float64(len(data))
	count := map[interface{}]int{}
	for _, item := range data {
		count[item] += 1
	}
	i := 0
	ps := make([]float64, len(count))
	for _, v := range count {
		ps[i] = float64(v) / total
		i++
	}
	return ps
}

func entropy(probs []float64) float64 {
	sum := 0.0
	for _, prob := range probs {
		if prob > 0 {
			sum -= prob * math.Log2(prob)
		}
	}
	return sum
}

func DataEntropy(data []interface{}) float64 {
	return entropy(clsprobs(data))
}

func PartitionBy(cands []*Candidate, attr string) Partition {
	partition := make(Partition)
	for _, cand := range cands {
		value := cand.Get(attr)
		if value != nil {
			partition[value] = append(partition[value], cand)
		}
	}
	return partition
}

func PartitionEntropyBy(partition Partition, attr string) float64 {
	sum := 0.0
	N := 0
	for _, subset := range partition {
		n := len(subset)
		N += n
		data := make([]interface{}, n)
		for i, cand := range subset {
			data[i] = cand.Get(attr)
		}
		sum += float64(n) * DataEntropy(data)
	}
	return sum / float64(N)
}

type Tree struct {
	attrib   string
	value    bool
	branches Branches
}

type Branches map[interface{}]*Tree

func New(attr string, val bool, b Branches) *Tree {
	return &Tree{attrib: attr, value: val, branches: b}
}

func (t *Tree) Classify(cand *Candidate) bool {
	if t.branches == nil {
		return t.value
	}
	if b, ok := t.branches[cand.Get(t.attrib)]; ok {
		return b.Classify(cand)
	} else {
		return t.value
	}
}

func BuildTree(cands []*Candidate, splitAttrs []string, targetAttr string) *Tree {
	labels := make(map[interface{}]int)
	for _, cand := range cands {
		val := cand.Get(targetAttr)
		labels[val] += 1
	}
	max := 0
	var mostCommon interface{}
	for label, count := range labels {
		if count > max {
			mostCommon = label
			max = count
		}
	}
	if len(labels) == 1 || len(splitAttrs) == 0 {
		return New("", mostCommon.(bool), nil)
	}
	var bestAttr string
	var bestPartition Partition
	min := 1.0
	for _, attr := range splitAttrs {
		partition := PartitionBy(cands, attr)
		imp := PartitionEntropyBy(partition, targetAttr)
		if imp < min {
			min = imp
			bestAttr = attr
			bestPartition = partition
		}
	}
	attrs := make([]string, 0, len(splitAttrs) - 1)
	for _, attr := range splitAttrs {
		if attr != bestAttr {
			attrs = append(attrs, attr)
		}
	}
	branches := make(Branches)
	for val, subset := range bestPartition {
		branches[val] = BuildTree(subset, attrs, targetAttr)
	}
	return New(bestAttr, mostCommon.(bool), branches)
}

File: pkg/tree/trees.py
import json
import csv
import dataclasses
import numpy as np
from functools import partial
from collections import Counter, defaultdict, namedtuple
from dataclasses import dataclass


@dataclass
class Candidate:
    level:  str
    lang:   str
    tweets: bool
    phd:    bool
    didwell: bool


# cands =  [
#     Candidate('Senior', 'Java',   False, False, False),
#     Candidate('Senior', 'Java',   False, True,  False),
#     Candidate('Mid',    'Python', False, False, True),
#     Candidate('Junior', 'Python', False, False, True),
#     Candidate('Junior', 'R',      True,  False, True),
#     Candidate('Junior', 'R',      True,  True,  False),
#     Candidate('Mid',    'R',      True,  True,  True),
#     Candidate('Senior', 'Python', False, False, False),
#     Candidate('Senior', 'R',      True,  False, True),
#     Candidate('Junior', 'Python', True,  False, True),
#     Candidate('Senior', 'Python', True,  True,  True),
#     Candidate('Mid',    'Python', False, True,  True),
#     Candidate('Mid',    'Java',   True,  False, True),
#     Candidate('Junior', 'Python', False, True,  False)
# ]

# def save(jsnfile):
#     with open(jsnfile, mode='w') as jout:
#         data = [dataclasses.asdict(c) for c in cands]
#         json.dump(data, jout, indent=4)

def fromJSON(jsnfile):
    with open(jsnfile) as jout:
        cands = json.load(jout)
        return [Candidate(**dic) for dic in cands]

def fromCSV(csvfile):
    cands = []
    with open(csvfile) as fin:
        _ = next(fin)
        for line in fin:
            rec = line.strip().split(',')
            cand = Candidate(*rec)
            cand.tweets = bool(int(cand.tweets))
            cand.phd = bool(int(cand.phd))
            cand.didwell = bool(int(cand.didwell))
            cands.append(cand)
    return cands

def fromCSV2(csvfile):
    cands = []
    with open(csvfile) as fin:
        rd = csv.reader(fin)
        _ = next(fin)
        for rec in rd:
            cand = Candidate(*rec)
            cand.tweets = bool(int(cand.tweets))
            cand.phd = bool(int(cand.phd))
            cand.didwell = bool(int(cand.didwell))
            cands.append(cand)
    return cands

def entropy(probs):
    '''stats: List[float]'''
    probs = np.array([p for p in probs if p > 0])
    return -(probs*np.log2(probs)).sum()

def clsprobs(data):
    '''data: List[float]'''
    total = len(data)
    return [c/total for c in Counter(data).values()]

def data_entropy(data):
    '''data: List[float]'''
    return entropy(clsprobs(data))

def partition_entropy(partition):
    '''partition: List[List[float]]'''
    total = sum(len(subset) for subset in partition)
    return sum([len(subset)*data_entropy(subset)/total for subset in partition])

def partition_by(cands, attr):
    '''cands: List[Candidate] attr: str'''
    partition = defaultdict(list)
    for cand in cands:
        value = getattr(cand, attr)
        partition[value].append(cand)
    return partition

def partition_entropy_by(partition, attr):
    label_partition = [
        [getattr(c, attr) for c in cands] 
        for cands in partition.values()
    ]
    return partition_entropy(label_partition)

Leaf = namedtuple('Leaf', 'value')
Split = namedtuple('Split', 'attr subtree value')

def classify(tree, cand):
    # check whether the answer can be given
    # base case of the recursion: tree is just a leaf
    if isinstance(tree, Leaf):
        return tree.value
    # get the value of the split attribute to figure out the next subtree
    branch = getattr(cand, tree.attr)
    # return default value if value of attribute is unknown to the tree
    if branch not in tree.subtree:
        return tree.value
    # delegate the classification task to the subtree
    subtree = tree.subtree[branch]
    return classify(subtree, cand)

def build_tree(cands, split_attrs, target_attr):
    '''ID3 algorithm'''
    # count target labels
    labels = Counter(getattr(cand, target_attr) for cand in cands)
    most_common = labels.most_common(1)[0][0]
    # if unique label or if there are no split attributes
    # then return majority label
    if len(labels) == 1 or not split_attrs:
        return Leaf(most_common)
    # otherwise split by the best attribute
    impurity = partial(split_entropy, cands, target_attr)
    best_attr = min(split_attrs, key=impurity)
    partition = partition_by(cands, best_attr)
    new_attrs = [a for a in split_attrs if a != best_attr]
    subtrees = {
        attr_value: build_tree(subset, new_attrs, target_attr)
        for attr_value, subset in partition.items()
    }
    return Split(best_attr, subtrees, most_common)

def split_entropy(cands, target_attr, attr):
    partition = partition_by(cands, attr)
    return partition_entropy_by(partition, target_attr)

# def classem(cands, tree):
#     c = 0
#     for cand in cands:
#         didwell = classify(tree, cand)
#         print(f'{cand.didwell} was classified as {didwell}')
#         if didwell == cand.didwell:
#             c += 1
#     n = len(cands)
#     r = c/n
#     print(f'Result: {c} correct of {n} (ratio: {r})')


if __name__ == '__main__':

    # save('cands.json')

    # cands = fromJSON('cands.json')
    cands = fromCSV2('cands.csv')
    
    cs = [c.phd for c in cands]
    x = data_entropy(cs)
    print(x)

    p = partition_by(cands, 'phd')
    q = partition_by(cands, 'lang')

    print('phd partition by phd', partition_entropy_by(p, 'phd'))
    print('phd partition by lang', partition_entropy_by(p, 'lang'))
    print('lang partition by phd', partition_entropy_by(q, 'phd'))
    print('lang partition by lang', partition_entropy_by(q, 'lang'))

    htree0 = Split(
        'level', {
            'Junior': Split('phd', {False: Leaf(True), True: Leaf(False)}, None),
            'Mid': Leaf(True),
            'Senior': Split('tweets', {False: Leaf(False), True: Leaf(True)}, None)
        },
        False
    )
    for cand in cands:
        print(f'Hire? : {cand.didwell} - {classify(htree0, cand)}')

    print("learned tree")
    # classem(cands, htree0)    

    htree = build_tree(cands, ['level', 'phd', 'tweets', 'lang'], 'didwell')
    for cand in cands:
        print(f'Hire? : {cand.didwell} - {classify(htree, cand)}')


    # classem(cands, htree)

    cand = Candidate('Intern', 'Java', True, True, None)
    ans = classify(htree, cand)
    print(f'Hire?: trained tree says {ans}')
    ans = classify(htree0, cand)
    print(f'Hire?: handcrafted tree says {ans}')
    
        
